tempesta_fw/http.c

```diff
@@ -2641,128 +2780,346 @@ tfw_http_hm_drop_resp(TfwHttpResp *resp)
 	tfw_http_msg_free((TfwHttpMsg *)req);
 }
 
+static void
+tfw_http_skb_queue(TfwHttpMsg *hm, struct sk_buff *skb,
+		   tfw_http_parse_stage_t stage)
+{
+	if (!tfw_http_msg_is_streamed(hm)) {
+		ss_skb_queue_tail(&hm->msg.head_skb, skb);
+		return;
+	}
+
+	spin_lock(&hm->msg.stream_lock);
+
+	switch (stage)
+	{
+	case TFW_HTTP_PARSE_HEADERS:
+		ss_skb_queue_tail(&hm->msg.head_skb, skb);
+		break;
+	case TFW_HTTP_PARSE_BODY:
+		ss_skb_queue_tail(&hm->msg.body_skb, skb);
+		break;
+	case TFW_HTTP_PARSE_TRAILER:
+		ss_skb_queue_tail(&hm->msg.trailer_skb, skb);
+		break;
+	default:
+		BUG();
+	}
+
+	spin_unlock(&hm->msg.stream_lock);
+}
+
+static int
+tfw_http_stream_req(TfwHttpReq *req)
+{
+	TfwSrvConn *conn = (TfwSrvConn *)req->state.stream_conn;
+	int r = 0;
+
+	TFW_DBG2("%s: resp=[%p]\n", __func__, req);
+
+	spin_lock(&conn->fwd_qlock);
+
+	if (!conn->msg_sent || ((TfwHttpReq *)conn->msg_sent != req))
+		goto err;
+	if (!tfw_srv_conn_get_if_live(conn))
+		goto err;
+	tfw_cli_conn_mod_katimer((TfwCliConn *)req->conn);
+	r = __tfw_http_conn_stream(req->state.stream_conn, (TfwHttpMsg *)req);
+	tfw_srv_conn_put(conn);
+err:
+	spin_unlock(&conn->fwd_qlock);
+
+	return r;
+}
+
+/**
+ * Parse and add a new skb into current message.
+ *
+ * Before the call @skb is not attached to the message. If parsing succeed,
+ * @skb is either attached to the message and @skb pointer is set to NULL;
+ * or it's split into two parts, first is attached to the message, second is
+ * returned in @skb pointer.
+ */
+static int
+tfw_http_parse_skb(TfwConn *conn, struct sk_buff **skb, unsigned int *off,
+		   TfwFsmData *data_up)
+{
+	TfwHttpMsg *hm = (TfwHttpMsg *)conn->msg;
+	TfwHttpParser *parser = &hm->parser;
+	bool client = TFW_CONN_TYPE(conn) & Conn_Clnt;
+	ss_skb_actor_t actor = client ? tfw_http_parse_req : tfw_http_parse_resp;
+	tfw_http_parse_stage_t stage;
+	unsigned int data_off;
+	size_t to_read;
+	int r = TFW_BLOCK;
+
+repeat:
+	if (likely(tfw_cfg_proxy_buff_sz > hm->msg.len))
+		to_read = tfw_cfg_proxy_buff_sz - hm->msg.len;
+	else
+		to_read = ULONG_MAX;
+
+	/*
+	 * Process/parse data in the SKB.
+	 * @data_off points at the start of data for processing.
+	 * @off is the current offset of data to process in
+	 * the SKB. After processing @off points at the end
+	 * of latest data chunk. However processing may have
+	 * stopped in the middle of the chunk. Adjust it to point
+	 * to the right location within the chunk.
+	 */
+	stage = tfw_http_parse_stage(hm);
+	parser->skb = *skb;
+	data_off = *off;
+	r = ss_skb_process(*skb, off, to_read, actor, hm);
```

Comment:
As I understand, there is no leakage due to `to_read` limitation since in current approach new skb (after splitting) is processed in the same `tfw_http_{req, resp}_process()` call and in the same manner as in sibling message case (but without creating a new message - new skb just parsed and added to current message on next iteration of https://github.com/tempesta-tech/tempesta/blob/7a2b3b7c74661a3799bbd3cbf36db3ca5a12a0f7/tempesta_fw/http.c#L3040 cycle). In current implementation of state machines in `tfw_http_{req, resp}_process()`, it seems that such new (`streamed`) skb is transmitted as soon as possible after it was created (in the same `tfw_http_{req, resp}_process()` call).
But I agree with point that it is better to handle buffering at lower layer also - to manage receive window size and ACK sending in order to limit transmission.

Reply:
Thank you guys for a review! I was thinking about this two  comments a lot and I came to conclusion that two approaches, low-level which is TCP receive window steering and high-level which is skb splitting, must be combined to effectively solve the #498 issue.

Receive window steering slow down talkative clients, which have requests pending in Tempesta's queues. With this option we control, how many memory is spent to store all client's requests. New requests from the client won't be ACKed and will stay in socket receive queue until memory consumption by the clients requests would go below buffering limit. The approach is significantly valuable for performance questions and must be implemented. But this approach  has nothing to do with original #498 which says, that Tempesta must forward messages by parts without fully assembling them. 

Intentional skb splitting in the middle of a message is only about proxying messages without assembling them inside Tempesta. The approach fixes only two very specific issue: don't assemble large messages inside Tempesta, forward them by parts as soon as a new part is received; and correctly serve long polling requests.

The main difference between approaches is a context: first one defines connection-wide limit, while the second one is message-wide.

I assume that the different views was reasoned by not very accurate `proxy_buffering` directive description. The directive shall be split into three parts:

1. `client_conn_bufferng N` - Controls receive window size steering for clients connections. `N` - size in bytes. This configuration option control haw many memory is used to store unanswered client requests. Newer requests are not acknowledged and are not removed from socket receive queue. The limit is applied to a client connection in whole.

2. `client_msg_buffering N` - Controls message buffering. Buffer only first N bytes of requests when forwarding requests to backend servers. Longer messages are forwarded part-by-part and never fully assembled in Tempesta. If request headers are longer than N bytes, then they are still buffered, since full set of headers is required to correctly serve and forward the request. The limit is applied per message and must not overcome  per-connecton `client_conn_bufferng` limit.

3. `server_msg_buffering N` - same as  `client_msg_buffering`  but for the server connections.

I've implemented points 2 and 3, point 1 is requested in comments above in this discussion.
@krizhanovsky @aleksostapenko , Any thoughts here? 

As you can see, there is no `server_conn_bufferng` limit in my proposal. I assume it to be dangerous and causing backed connections deadlocks. Let's imagine that the first request from all clients is long polling request or very heavy request required significant time to complete. Other pipelined requests from that clients are plain requests easy to complete. While some server connections are stuck in serving heavy/long-polling requests, others serving the rest of requests. With `server_conn_bufferng` limit second group of server connections will be blocked until heavy/long-polling requests are replied. From now on new client connections can't be served for undefined time: all server connections are blocked.

Reply:
> As I understand, there is no leakage due to...

Yes, @aleksostapenko  you're got the idea right.

Reply:
The main point for #498 is not to limit memory spent for HTTP messages handling. If we took an skb from the client connection, it's not larger than 64KB, then why not to fully process it? There is no reason to postpone it in some queue or call our HTTP processing logic twice, instead we should process as much as possible in as less shots as possible. My comment was about we must make a client not to send us more skbs to process. But all skbs received by the TCP/IP stack must be processed immediately.

Reply:
Actually, this is the 3rd point of #391 with the only difference that sent receive window is controlled by the buffer size in this case and by QoS in #488 case, so the window control logic shall be generic enough for #488.